examples/fit_with_bounds.py

```diff
@@ -18,78 +18,92 @@
+        variable_symbols = axis_names & all_symbols
+        parameter_symbols = all_symbols - variable_symbols
+        variable_symbols = tuple(variable_symbols)
+        variable_names = tuple([str(j) for j in variable_symbols])
+        parameter_symbols = tuple(parameter_symbols)
+        parameter_names = tuple([str(j) for j in parameter_symbols])
+        logger.debug(
+            strm(
+                "all symbols are",
+                all_symbols,
+                "axis names are",
+                axis_names,
+                "variable names are",
+                variable_names,
+                "parameter names are",
+                parameter_names,
+            )
         )
-    )
-    # }}}
-    pars = Parameters()
-    for this_name in parameter_names:
-        kwargs = {}
-        if this_name in guesses.keys():
-            logger.info(strm("applying bounds for", this_name))
-            kwargs.update(guesses[str(this_name)])
-        pars.add(this_name, **kwargs)
-    for j in pars:
-        logger.info(strm("fit param ---", j))
-    logger.info(strm(pars))
-    fn = lambdify(
-        variable_symbols + parameter_symbols,
-        expr,
-        modules=[{"ImmutableMatrix": np.ndarray}, "numpy", "scipy"],
-    )
-    return pars, fn
-def residual(pars, x, data=None):
-    "calculate the residual OR if data is None, return fake data"
-    shift = pars["shift"]
-    if abs(shift) > pi / 2:
-        shift = shift - sign(shift) * pi
-    model = myfunc(x, **pars.valuesdict())
-    if data is None:
-        return model
-    return model - data
+        # }}}
+        pars = Parameters()
+        for this_name in parameter_names:
+            kwargs = {}
+            if this_name in guesses.keys():
+                logger.debug(strm("applying bounds for", this_name))
+                kwargs.update(guesses[str(this_name)])
+            pars.add(this_name, **kwargs)
+        for j in pars:
+            logger.info(strm("fit param ---", j))
+        logger.info(strm(pars))
+        self.fn = lambdify(
+            variable_symbols + parameter_symbols,
+            expr,
+            modules=[{"ImmutableMatrix": np.ndarray}, "numpy", "scipy"],
+        )
+        return pars
+
+
+    def residual(self, pars, x, data=None):
+        "calculate the residual OR if data is None, return fake data"
+        shift = pars["shift"]
+        if abs(shift) > pi / 2:
+            shift = shift - sign(shift) * pi
+        model = self.fn(x, **pars.valuesdict())
+        if data is None:
+            return model
+        return model - data
+
+
+# }}}
+thisfit = myfitclass()
+# {{{ a lot of what's below depends on knowing what the shape and dimension labels of my data are, so define that here
+x_vals = linspace(0, 250, 1500)
+empty_data = nddata(x_vals, "x").copy(data=False)
 # }}}
 # {{{making sympy expression
 A, shift, period, decay, x = sp.symbols("A shift period decay x")
 expr = A * sp.sin(shift + x / period) * sp.exp(-((x * decay) ** 2))
```

Comment:
so -- as an exercise, maybe try to change this: rather than defining as a global variable and passing to the function below, can you rather set an attribute of `thisfit` here, an then change `gen_from_expr` so that it doesn't require an expression argument?

Reply:
make the expression an attribute of `thisfit`? but wouldn't that not be a good idea if we wanted to fit to a different expression it would be hardcoded in the class?